### PR TITLE
A20, Task 1 & 2

### PR DESCRIPTION
**Pull request by vikrampg1710-tmm**
Tried Tasks 1 and 2.
(Task 3 not completed).

### CODE REVIEW

File: Parser/Types/TypeAnalyze.cs [hunk 1]
```diff
@@ -26,14 +28,24 @@ public override NType Visit (NProgram p)
...
+
    public override NType Visit (NVarDecl d) {
       mSymbols.Vars.Add (d);
       return d.Type;
    }
 
    public override NType Visit (NFnDecl f) {
       mSymbols.Funcs.Add (f);
-      return f.Return;
+      if (mSymbols.Find (f.Name.Text) is not NVarDecl) {
```
Comment (tarydon) [inline Parser/Types/TypeAnalyze.cs]: What is this if check for (not NVarDecl). Isn't it already known that f is a function?

File: Parser/Types/TypeAnalyze.cs [hunk 2]
```diff
@@ -72,9 +84,10 @@ public override NType Visit (NWriteStmt w)
       f.Start.Accept (this); f.End.Accept (this); f.Body.Accept (this);
       return Void;
    }
-
+   
    public override NType Visit (NReadStmt r) {
-      throw new NotImplementedException ();
+      Visit ((IEnumerable<Node>)r.Vars[0]);
```
Comment (tarydon) [inline Parser/Types/TypeAnalyze.cs]: r.Vars[0] will never be of type (IEnumerable<Node>). Did you check if this works?

File: Parser/Types/TypeAnalyze.cs [hunk 3]
```diff
@@ -86,9 +99,11 @@ public override NType Visit (NWriteStmt w)
       Visit (r.Stmts); r.Condition.Accept (this);
       return Void;
    }
-
+    
    public override NType Visit (NCallStmt c) {
-      throw new NotImplementedException ();
+      if (mSymbols.Find (c.Name.Text) is NVarDecl)
+         Visit (c.Params);
+      throw new ParseException (c.Name, "Unknown variable");
```
Comment (tarydon) [inline Parser/Types/TypeAnalyze.cs]: You should expect to find a NFnDecl, not an NVarDecl here. Did you test this?

File: Parser/Types/TypeAnalyze.cs [hunk 4]
```diff
@@ -138,9 +153,11 @@ public override NType Visit (NUnary u)
          return d.Type = v.Type;
       throw new ParseException (d.Name, "Unknown variable");
    }
-
+  
    public override NType Visit (NFnCall f) {
-      throw new NotImplementedException ();
+      if (mSymbols.Find (f.Name.Text) is not NVarDecl)
```
Comment (tarydon) [inline Parser/Types/TypeAnalyze.cs]: Again, I am not sure why this if statement is needed here.

File: Parser/Util/PSIPrint.cs [hunk 5]
```diff
@@ -24,6 +24,12 @@ public override StringBuilder Visit (NBlock b)
       return S;
    }
 
+   public override StringBuilder Visit (NConstDecl c) {
+      NWrite ("const"); N++;
+      NWrite ($"{c.Name} = {c.Value}");
+      return S;
```
Comment (tarydon) [inline Parser/Util/PSIPrint.cs]: If you have multiple **const** declarations like this:
```
const
  pi = 3.14;
  message = "Hello";
  n = 4;
```
Will you program print just one **const** keyword, or will it print it 3 times? Test.

**Review by tarydon — CHANGES_REQUESTED** [review]

File: Parser/Types/TypeAnalyze.cs (hunk 2 above)
Comment (vikrampg1710-tmm) [reply Parser/Types/TypeAnalyze.cs]: It's not working.  I just used the intellisense hints.

File: Parser/Types/TypeAnalyze.cs (hunk 3 above)
Comment (vikrampg1710-tmm) [reply Parser/Types/TypeAnalyze.cs]: The compilation is crashing.
Is Fncall different from CallStmt?

**Pull request closed without merging**